collector/physical_disk.go

```diff
@@ -0,0 +1,269 @@
+			"Seconds that the disk was idle (PhysicalDisk.PercentIdleTime)",
+			[]string{"disk"},
+			nil,
+		),
+
+		SplitIOs: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "split_ios_total"),
+			"The number of I/Os to the disk were split into multiple I/Os (PhysicalDisk.SplitIOPerSec)",
+			[]string{"disk"},
+			nil,
+		),
+
+		ReadLatency: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "read_latency_seconds_total"),
+			"Shows the average time, in seconds, of a read operation from the disk (PhysicalDisk.AvgDiskSecPerRead)",
+			[]string{"disk"},
+			nil,
+		),
+
+		WriteLatency: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "write_latency_seconds_total"),
+			"Shows the average time, in seconds, of a write operation to the disk (PhysicalDisk.AvgDiskSecPerWrite)",
+			[]string{"disk"},
+			nil,
+		),
+
+		ReadWriteLatency: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "read_write_latency_seconds_total"),
+			"Shows the time, in seconds, of the average disk transfer (PhysicalDisk.AvgDiskSecPerTransfer)",
+			[]string{"disk"},
+			nil,
+		),
+
+		diskWhitelistPattern: regexp.MustCompile(fmt.Sprintf("^(?:%s)$", *diskWhitelist)),
+		diskBlacklistPattern: regexp.MustCompile(fmt.Sprintf("^(?:%s)$", *diskBlacklist)),
+	}, nil
+}
+
+// Collect sends the metric values for each metric
+// to the provided prometheus Metric channel.
+func (c *PhysicalDiskCollector) Collect(ctx *ScrapeContext, ch chan<- prometheus.Metric) error {
+	if desc, err := c.collect(ctx, ch); err != nil {
+		log.Error("failed collecting physical_disk metrics:", desc, err)
+		return err
+	}
+	return nil
+}
+
+// Win32_PerfRawData_PerfDisk_PhysicalDisk docs:
+// - https://docs.microsoft.com/en-us/previous-versions/aa394308(v=vs.85) - Win32_PerfRawData_PerfDisk_PhysicalDisk class
+type PhysicalDisk struct {
+	Name                   string
+	CurrentDiskQueueLength float64 `perflib:"Current Disk Queue Length"`
+	DiskReadBytesPerSec    float64 `perflib:"Disk Read Bytes/sec"`
+	DiskReadsPerSec        float64 `perflib:"Disk Reads/sec"`
+	DiskWriteBytesPerSec   float64 `perflib:"Disk Write Bytes/sec"`
+	DiskWritesPerSec       float64 `perflib:"Disk Writes/sec"`
+	PercentDiskReadTime    float64 `perflib:"% Disk Read Time"`
+	PercentDiskWriteTime   float64 `perflib:"% Disk Write Time"`
+	PercentIdleTime        float64 `perflib:"% Idle Time"`
+	SplitIOPerSec          float64 `perflib:"Split IO/Sec"`
+	AvgDiskSecPerRead      float64 `perflib:"Avg. Disk sec/Read"`
+	AvgDiskSecPerWrite     float64 `perflib:"Avg. Disk sec/Write"`
+	AvgDiskSecPerTransfer  float64 `perflib:"Avg. Disk sec/Transfer"`
+}
+
+func (c *PhysicalDiskCollector) collect(ctx *ScrapeContext, ch chan<- prometheus.Metric) (*prometheus.Desc, error) {
+	var dst []PhysicalDisk
+	if err := unmarshalObject(ctx.perfObjects["PhysicalDisk"], &dst); err != nil {
+		return nil, err
+	}
+
+	for _, disk := range dst {
```

Comment:
The `disk` label currently looks like this: `"0 C:"`. Since I remember your use-case was in non-mounted disks, what does it look like under those circumstances @rob-scheepens?
I'm not super happy about having both disk number **and** mountpoint, to be honest. Depending on what the label is for non-mounted, perhaps we can do something that'll look reasonable for both cases.

Reply:
For non-mounted disks you will only see the disk number. I agree not to have the label and number, I was actually surprised to see C: show up. ;)

Reply:
Was this fixed?

Reply:
Nope, not yet.

Reply:
I'd like this to be resolved before we merge, since once this is released it's breaking to change the labels. I can see three different cases that can occur, maybe missing some?
- Not mounted
- Mounted as disk root
- Mounted as a directory inside another (eg D:\Foo\Bar)

Some possible resolutions:
- Always use disk number, skip the mountpoint
- Use mountpoint if available (I dislike this, since it'd mean you'd get a label change by mounting the disk)
- Label with disk number as `disk`, create a separate `_info` metric which has the mountpoint

What do you think?

Reply:
I'd vote for a fix that shows Raw as label for disks that do not have a partition, and thus are also not mounted. We could combine this with the option of always using disk number, so that in case of a raw disk we would see "1 Raw", and in the case of a mounted partition, we would see the behavior as it currently is. Does that make sense?

Reply:
Testing Labels
==============
Test vm's disks:
```
DISKPART> lis dis

  Disk ###  Status         Size     Free     Dyn  Gpt
  --------  -------------  -------  -------  ---  ---
  Disk 0    Online          100 GB      0 B        *
  Disk 1    Offline          20 GB    20 GB
  Disk 2    Offline          20 GB    20 GB
  Disk 3    Offline          20 GB    20 GB
  Disk 4    Offline          20 GB    20 GB
  Disk 5    Offline          20 GB    20 GB
  Disk 6    Offline          20 GB    20 GB
  Disk 7    Online           50 GB    50 GB
  Disk 8    Online           50 GB    50 GB

windows_physical_disk_idle_seconds_total{disk="1"} 1742.6185977999999
windows_physical_disk_idle_seconds_total{disk="2"} 1742.6185988
windows_physical_disk_idle_seconds_total{disk="3"} 1742.6185944
windows_physical_disk_idle_seconds_total{disk="4"} 1742.6185948999998
windows_physical_disk_idle_seconds_total{disk="5"} 1742.6185957999999
windows_physical_disk_idle_seconds_total{disk="6"} 1742.6185965
windows_physical_disk_idle_seconds_total{disk="7"} 40.4170576
windows_physical_disk_idle_seconds_total{disk="8"} 27.7063651
```
Note: disk 0 is excluded by blacklisting C:. None of disks 1-8 contain any partition.

Created partition on disk 7 using "crea par pri". No change in labels:
```
windows_physical_disk_idle_seconds_total{disk="1"} 1871.906501
windows_physical_disk_idle_seconds_total{disk="2"} 1871.9065021
windows_physical_disk_idle_seconds_total{disk="3"} 1871.9064982
windows_physical_disk_idle_seconds_total{disk="4"} 1871.9064985999998
windows_physical_disk_idle_seconds_total{disk="5"} 1871.9064994999999
windows_physical_disk_idle_seconds_total{disk="6"} 1871.9065008
windows_physical_disk_idle_seconds_total{disk="7"} 169.68967189999998
windows_physical_disk_idle_seconds_total{disk="8"} 156.99427
```
Assigned letter X using "assign letter=x", no change in labels still. 
Also after formatting no change in labels. Only after reboot the X: volume shows up in the metrics list:
```
windows_physical_disk_idle_seconds_total{disk="1"} 7.8161184
windows_physical_disk_idle_seconds_total{disk="2"} 7.8161192999999995
windows_physical_disk_idle_seconds_total{disk="3"} 7.8161203
windows_physical_disk_idle_seconds_total{disk="4"} 7.8161213
windows_physical_disk_idle_seconds_total{disk="5"} 7.8161219
windows_physical_disk_idle_seconds_total{disk="6"} 7.8161236999999995
windows_physical_disk_idle_seconds_total{disk="7 X:"} 7.816124599999999
windows_physical_disk_idle_seconds_total{disk="8"} 7.8161261
```
Is this an issue with the exporter? The interesting thing is that when I did the same workflow for Z: on disk 8 (crea par pri, format ntfs quick, assign letter z), it did show up immediately:
```
windows_physical_disk_idle_seconds_total{disk="1"} 0.00039109999999999997
windows_physical_disk_idle_seconds_total{disk="2"} 0.00039139999999999997
windows_physical_disk_idle_seconds_total{disk="3"} 0.0003917
windows_physical_disk_idle_seconds_total{disk="4"} 0.0003915
windows_physical_disk_idle_seconds_total{disk="5"} 0.00039139999999999997
windows_physical_disk_idle_seconds_total{disk="6"} 0.0003915
windows_physical_disk_idle_seconds_total{disk="7 X:"} 0.00039139999999999997
windows_physical_disk_idle_seconds_total{disk="8 Z:"} 0.0003917
```
Deleting the partion was also immediately reflected:
```
windows_physical_disk_idle_seconds_total{disk="1"} 90.2290264
windows_physical_disk_idle_seconds_total{disk="2"} 90.2290273
windows_physical_disk_idle_seconds_total{disk="3"} 90.2290281
windows_physical_disk_idle_seconds_total{disk="4"} 90.22902839999999
windows_physical_disk_idle_seconds_total{disk="5"} 90.229029
windows_physical_disk_idle_seconds_total{disk="6"} 90.2290299
windows_physical_disk_idle_seconds_total{disk="7 X:"} 90.2290305
windows_physical_disk_idle_seconds_total{disk="8"} 90.1772701
```
Creating two partitions on disk 8 also was immediately reflected:
```
windows_physical_disk_idle_seconds_total{disk="1"} 157.879834
windows_physical_disk_idle_seconds_total{disk="2"} 157.8798352
windows_physical_disk_idle_seconds_total{disk="3"} 157.8798363
windows_physical_disk_idle_seconds_total{disk="4"} 157.8798369
windows_physical_disk_idle_seconds_total{disk="5"} 157.87983739999999
windows_physical_disk_idle_seconds_total{disk="6"} 157.8798382
windows_physical_disk_idle_seconds_total{disk="7 X:"} 157.8798388
windows_physical_disk_idle_seconds_total{disk="8 Y: Z:"} 157.77886379999998
```
To me, this labeling looks consistent and in line with what one would expact. You agree @carlpett? If so, we can put this conversation to rest and merge the PR, right?

Reply:
From a brief review of this conversation I assume option 2 has been chosen?

> Use mountpoint if available (I dislike this, since it'd mean you'd get a label change by mounting the disk)

If so I'd recommend documenting this labeling behaviour, in order to avoid user surprise. Could you add an entry for this collector in the `docs/` directory?

Reply:
As mentioned in https://github.com/prometheus-community/windows_exporter/pull/803#issuecomment-1198869937, I'd like the behavior where the disk number is always shown, whether or not in combination with a partition letter or mount point, in case there is a mounted partition. If you agree, let me know and I'll create the doc.

Reply:
I think either the disk number only or having the disk number and drive letter as separate labels would be fine.
E.G. `windows_physical_disk_idle_seconds_total{disk="7", drive_letter="X:"} 157.8798388`

Keeping the two pieces of information in a single label would have users resorting to regexes for label matching: `windows_physical_disk_idle_seconds_total{disk=".*X:"}` or `windows_physical_disk_idle_seconds_total{disk="7.*"}`